File: app/src/instructions/createProposalPartTwo.ts
import { createAssociatedTokenAccountInstruction } from "@solana/spl-token";
import {AutocratClient} from "../AutocratClient";
import {InstructionHandler} from "../InstructionHandler";
import { getATA, getConditionalOnFailMetaMintAddr, getConditionalOnFailUsdcMintAddr, getConditionalOnPassMetaMintAddr, getConditionalOnPassUsdcMintAddr, getDaoAddr, getDaoTreasuryAddr, getFailMarketAmmAddr, getPassMarketAmmAddr, getProposalAddr, getProposalInstructionsAddr } from '../utils';
import BN from "bn.js";

export const createProposalPartTwoHandler = async (
    client: AutocratClient,
    initialPassMarketPriceUnits: number,
    initialFailMarketPriceUnits: number,
    quoteLiquidityAtomsPerAmm: BN,
): Promise<InstructionHandler> => {
    let dao = await client.program.account.dao.fetch(getDaoAddr(client.program.programId)[0])

    let proposalNumber = dao.proposalCount - 1

    let proposalAddr = getProposalAddr(client.program.programId, proposalNumber)[0]

    let conditionalOnPassMetaMint = getConditionalOnPassMetaMintAddr(client.program.programId, proposalNumber)[0]
    let conditionalOnPassUsdcMint = getConditionalOnPassUsdcMintAddr(client.program.programId, proposalNumber)[0]
    let conditionalOnFailMetaMint = getConditionalOnFailMetaMintAddr(client.program.programId, proposalNumber)[0]
    let conditionalOnFailUsdcMint = getConditionalOnFailUsdcMintAddr(client.program.programId, proposalNumber)[0]

    let ix = await client.program.methods
        .createProposalPartTwo(
            initialPassMarketPriceUnits,
            initialFailMarketPriceUnits,
            quoteLiquidityAtomsPerAmm
        )
        .accounts({
            proposer: client.provider.publicKey,
            proposal: proposalAddr,
            passMarketAmm: getPassMarketAmmAddr(client.program.programId, proposalNumber)[0],
            failMarketAmm: getFailMarketAmmAddr(client.program.programId, proposalNumber)[0],
            metaMint: dao.metaMint,
            usdcMint: dao.usdcMint,
            conditionalOnPassMetaMint,
            conditionalOnPassUsdcMint,
            conditionalOnFailMetaMint,
            conditionalOnFailUsdcMint,
            metaProposerAta: getATA(dao.metaMint, client.provider.publicKey)[0],
            usdcProposerAta: getATA(dao.usdcMint, client.provider.publicKey)[0],
            metaVaultAta: getATA(dao.metaMint, proposalAddr)[0],
            usdcVaultAta: getATA(dao.usdcMint, proposalAddr)[0],
            conditionalOnPassMetaVaultAta: getATA(conditionalOnPassMetaMint, proposalAddr)[0],
            conditionalOnPassUsdcVaultAta: getATA(conditionalOnPassUsdcMint, proposalAddr)[0],
            conditionalOnFailMetaVaultAta: getATA(conditionalOnFailMetaMint, proposalAddr)[0],
            conditionalOnFailUsdcVaultAta: getATA(conditionalOnFailUsdcMint, proposalAddr)[0],
        })
        .instruction()

    let metaVaultATA = getATA(dao.metaMint, proposalAddr)[0]
    let usdcVaultATA = getATA(dao.usdcMint, proposalAddr)[0]

    let metaAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        metaVaultATA,
        proposalAddr,
        dao.metaMint,
    )

    let usdcAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        usdcVaultATA,
        proposalAddr,
        dao.usdcMint,
    )

    let conditionalOnPassMetaVaultATA = getATA(conditionalOnPassMetaMint, proposalAddr)[0]
    let conditionalOnPassUsdcVaultATA = getATA(conditionalOnPassUsdcMint, proposalAddr)[0]
    let conditionalOnFailMetaVaultATA = getATA(conditionalOnFailMetaMint, proposalAddr)[0]
    let conditionalOnFailUsdcVaultATA = getATA(conditionalOnFailUsdcMint, proposalAddr)[0]

    let passMetaAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        conditionalOnPassMetaVaultATA,
        proposalAddr,
        conditionalOnPassMetaMint,
    )

    let passUsdcAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        conditionalOnPassUsdcVaultATA,
        proposalAddr,
        conditionalOnPassUsdcMint,
    )

    let failMetaAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        conditionalOnFailMetaVaultATA,
        proposalAddr,
        conditionalOnFailMetaMint,
    )

    let failUsdcAtaIx = createAssociatedTokenAccountInstruction(
        client.provider.publicKey,
        conditionalOnFailUsdcVaultATA,
        proposalAddr,
        conditionalOnFailUsdcMint,
    )
   
    return new InstructionHandler([metaAtaIx, usdcAtaIx, passMetaAtaIx, passUsdcAtaIx, failMetaAtaIx, failUsdcAtaIx, ix], [], client)
};

File: programs/autocrat/src/instructions/autocrat/create_proposal_part_two.rs
use anchor_lang::prelude::*;
use anchor_lang::solana_program;
use anchor_spl::associated_token;
use anchor_spl::associated_token::AssociatedToken;
use anchor_spl::token;
use anchor_spl::token::Mint;
use anchor_spl::token::Token;
use anchor_spl::token::TokenAccount;

use crate::error::ErrorCode;
use crate::state::*;

#[derive(Accounts)]
pub struct CreateProposalPartTwo<'info> {
    #[account(mut)]
    pub proposer: Signer<'info>,
    #[account(
        mut,
        seeds = [
            b"proposal",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub proposal: Box<Account<'info, Proposal>>,
    #[account(
        mut,
        seeds = [
            b"pass_market_amm",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub pass_market_amm: Box<Account<'info, Amm>>,
    #[account(
        mut,
        seeds = [
            b"fail_market_amm",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub fail_market_amm: Box<Account<'info, Amm>>,
    #[account(
        constraint = meta_mint.key() == proposal.meta_mint.key()
    )]
    pub meta_mint: Box<Account<'info, Mint>>,
    #[account(
        constraint = usdc_mint.key() == proposal.usdc_mint.key()
    )]
    pub usdc_mint: Box<Account<'info, Mint>>,
    #[account(
        mut,
        seeds = [
            b"conditional_on_pass_meta",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub conditional_on_pass_meta_mint: Box<Account<'info, Mint>>,
    #[account(
        mut,
        seeds = [
            b"conditional_on_pass_usdc",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub conditional_on_pass_usdc_mint: Box<Account<'info, Mint>>,
    #[account(
        mut,
        seeds = [
            b"conditional_on_fail_meta",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub conditional_on_fail_meta_mint: Box<Account<'info, Mint>>,
    #[account(
        mut,
        seeds = [
            b"conditional_on_fail_usdc",
            proposal.number.to_le_bytes().as_ref(),
        ],
        bump
    )]
    pub conditional_on_fail_usdc_mint: Box<Account<'info, Mint>>,
    #[account(
        mut,
        associated_token::mint = meta_mint,
        associated_token::authority = proposer,
    )]
    pub meta_proposer_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = usdc_mint,
        associated_token::authority = proposer,
    )]
    pub usdc_proposer_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = meta_mint,
        associated_token::authority = proposal,
    )]
    pub meta_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = usdc_mint,
        associated_token::authority = proposal,
    )]
    pub usdc_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = conditional_on_pass_meta_mint,
        associated_token::authority = proposal,
    )]
    pub conditional_on_pass_meta_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = conditional_on_pass_usdc_mint,
        associated_token::authority = proposal,
    )]
    pub conditional_on_pass_usdc_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = conditional_on_fail_meta_mint,
        associated_token::authority = proposal,
    )]
    pub conditional_on_fail_meta_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(
        mut,
        associated_token::mint = conditional_on_fail_usdc_mint,
        associated_token::authority = proposal,
    )]
    pub conditional_on_fail_usdc_vault_ata: Box<Account<'info, TokenAccount>>,
    #[account(address = associated_token::ID)]
    pub associated_token_program: Program<'info, AssociatedToken>,
    #[account(address = token::ID)]
    pub token_program: Program<'info, Token>,
    pub system_program: Program<'info, System>,
}

pub fn handler(
    ctx: Context<CreateProposalPartTwo>,
    initial_pass_market_price_units: f32, // human-readable price (i.e. units)
    initial_fail_market_price_units: f32, // human-readable price (i.e. units)
    quote_liquidity_atoms_per_amm: u64,
) -> Result<()> {
    let CreateProposalPartTwo {
        proposer,
        proposal,
        pass_market_amm,
        fail_market_amm,
        meta_mint,
        usdc_mint,
        conditional_on_pass_meta_mint,
        conditional_on_pass_usdc_mint,
        conditional_on_fail_meta_mint,
        conditional_on_fail_usdc_mint,
        meta_proposer_ata,
        usdc_proposer_ata,
        meta_vault_ata,
        usdc_vault_ata,
        conditional_on_pass_meta_vault_ata,
        conditional_on_pass_usdc_vault_ata,
        conditional_on_fail_meta_vault_ata,
        conditional_on_fail_usdc_vault_ata,
        associated_token_program: _,
        token_program: _,
        system_program: _,
    } = ctx.accounts;

    assert!(proposal.part_one_complete);
    assert!(!proposal.part_two_complete);
    proposal.part_two_complete = true;

    let clock = Clock::get()?;
    proposal.slot_enqueued = clock.slot;

    pass_market_amm.ltwap_slot_updated = clock.slot;
    fail_market_amm.ltwap_slot_updated = clock.slot;

    // ==== deposit initial liquidity ====
    // TODO

    Ok(())
}
